File: core/src/sessions/gemini.rs
//! Gemini CLI session parser
//!
//! Parses JSON session files from ~/.gemini/tmp/*/chats/session-*.json

use std::path::Path;
use serde::Deserialize;
use crate::TokenBreakdown;
use super::UnifiedMessage;

/// Gemini session structure
#[derive(Debug, Deserialize)]
pub struct GeminiSession {
    #[serde(rename = "sessionId")]
    pub session_id: String,
    #[serde(rename = "projectHash")]
    pub project_hash: String,
    #[serde(rename = "startTime")]
    pub start_time: String,
    #[serde(rename = "lastUpdated")]
    pub last_updated: String,
    pub messages: Vec<GeminiMessage>,
}

/// Gemini message structure
#[derive(Debug, Deserialize)]
pub struct GeminiMessage {
    pub id: String,
    pub timestamp: Option<String>,
    #[serde(rename = "type")]
    pub message_type: String,
    pub content: Option<String>,
    pub tokens: Option<GeminiTokens>,
    pub model: Option<String>,
}

/// Gemini token structure
#[derive(Debug, Deserialize)]
pub struct GeminiTokens {
    pub input: Option<i64>,
    pub output: Option<i64>,
    pub cached: Option<i64>,
    pub thoughts: Option<i64>,
    pub tool: Option<i64>,
    pub total: Option<i64>,
}

/// Parse a Gemini session file
pub fn parse_gemini_file(path: &Path) -> Vec<UnifiedMessage> {
    let data = match std::fs::read(path) {
        Ok(d) => d,
        Err(_) => return Vec::new(),
    };
    
    let mut bytes = data;
    let session: GeminiSession = match simd_json::from_slice(&mut bytes) {
        Ok(s) => s,
        Err(_) => return Vec::new(),
    };
    
    let mut messages = Vec::new();
    
    for msg in session.messages {
        // Only process gemini messages with token data
        if msg.message_type != "gemini" {
            continue;
        }
        
        let tokens = match msg.tokens {
            Some(t) => t,
            None => continue,
        };
        
        let model = match msg.model {
            Some(m) => m,
            None => continue,
        };
        
        let timestamp = msg.timestamp
            .and_then(|ts| chrono::DateTime::parse_from_rfc3339(&ts).ok())
            .map(|dt| dt.timestamp_millis())
            .unwrap_or(0);
        
        if timestamp == 0 {
            continue;
        }
        
        messages.push(UnifiedMessage::new(
            "gemini",
            model,
            "google",
            timestamp,
            TokenBreakdown {
                input: tokens.input.unwrap_or(0),
                output: tokens.output.unwrap_or(0),
                cache_read: tokens.cached.unwrap_or(0),
                cache_write: 0,
                reasoning: tokens.thoughts.unwrap_or(0),
            },
            0.0, // Cost calculated later
        ));
    }
    
    messages
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_gemini_structure() {
        let json = r#"{
            "sessionId": "ses_123",
            "projectHash": "abc123",
            "startTime": "2025-06-15T12:00:00Z",
            "lastUpdated": "2025-06-15T12:30:00Z",
            "messages": [
                {
                    "id": "msg_1",
                    "timestamp": "2025-06-15T12:00:00Z",
                    "type": "user",
                    "content": "Hello"
                },
                {
                    "id": "msg_2",
                    "timestamp": "2025-06-15T12:01:00Z",
                    "type": "gemini",
                    "content": "Hi there!",
                    "model": "gemini-2.0-flash",
                    "tokens": {
                        "input": 10,
                        "output": 20,
                        "cached": 5,
                        "thoughts": 0,
                        "tool": 0,
                        "total": 35
                    }
                }
            ]
        }"#;

        let mut bytes = json.as_bytes().to_vec();
        let session: GeminiSession = simd_json::from_slice(&mut bytes).unwrap();
        
        assert_eq!(session.messages.len(), 2);
        assert_eq!(session.messages[1].model, Some("gemini-2.0-flash".to_string()));
    }
}

File: packages/cli/src/sessions/gemini.ts
/**
 * Gemini CLI session parser
 * Reads from ~/.gemini/tmp/{projectHash}/chats/session-*.json
 */

import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { createUnifiedMessage, type UnifiedMessage, type TokenBreakdown } from "./types.js";

interface GeminiMessage {
  id: string;
  timestamp: string;
  type: "user" | "gemini";
  content: string;
  tokens?: {
    input: number;
    output: number;
    cached: number;
    thoughts: number;
    tool: number;
    total: number;
  };
  model?: string;
}

interface GeminiSession {
  sessionId: string;
  projectHash: string;
  startTime: string;
  lastUpdated: string;
  messages: GeminiMessage[];
}

export function getGeminiBasePath(): string {
  return path.join(os.homedir(), ".gemini");
}

export function parseGeminiMessages(): UnifiedMessage[] {
  const basePath = getGeminiBasePath();
  const tmpPath = path.join(basePath, "tmp");

  if (!fs.existsSync(tmpPath)) {
    return [];
  }

  const messages: UnifiedMessage[] = [];

  try {
    // Find all project directories
    const projectDirs = fs
      .readdirSync(tmpPath, { withFileTypes: true })
      .filter((d) => d.isDirectory())
      .map((d) => path.join(tmpPath, d.name));

    for (const projectDir of projectDirs) {
      const chatsDir = path.join(projectDir, "chats");
      if (!fs.existsSync(chatsDir)) continue;

      // Find all session JSON files
      const sessionFiles = fs
        .readdirSync(chatsDir)
        .filter((f) => f.startsWith("session-") && f.endsWith(".json"));

      for (const sessionFile of sessionFiles) {
        try {
          const content = fs.readFileSync(path.join(chatsDir, sessionFile), "utf-8");
          const session = JSON.parse(content) as GeminiSession;

          for (const msg of session.messages) {
            // Only process gemini messages with token data and timestamp
            if (msg.type !== "gemini" || !msg.tokens || !msg.model || !msg.timestamp) continue;

            const timestamp = new Date(msg.timestamp).getTime();

            // Skip invalid timestamps
            if (isNaN(timestamp)) continue;

            const tokens: TokenBreakdown = {
              input: msg.tokens.input || 0,
              output: msg.tokens.output || 0,
              cacheRead: msg.tokens.cached || 0,
              cacheWrite: 0, // Gemini doesn't track cache write
              reasoning: msg.tokens.thoughts || 0,
            };

            messages.push(
              createUnifiedMessage(
                "gemini",
                msg.model,
                "google",
                session.sessionId,
                timestamp,
                tokens
              )
            );
          }
        } catch {
          // Skip malformed files
        }
      }
    }
  } catch {
    // Skip inaccessible directories
  }

  return messages;
}
